File: leetcode-04/solutions/cpp/0093.cpp
class Solution {
 public:
  vector<string> restoreIpAddresses(string s) {
    vector<string> ans;
    vector<string> path(4);

    dfs(s, 0, 0, path, ans);

    return ans;
  }

  void dfs(string s, int depth, int index, vector<string>& path,
           vector<string>& ans) {
    if (depth == 4 && index == s.length()) {
      ans.push_back(path[0] + "." + path[1] + "." + path[2] + "." + path[3]);
      return;
    }
    if (depth == 4 || index == s.length()) return;

    for (int i = 1; i <= 3; ++i) {
      if (index + i > s.length()) return;
      if (i > 1 && s[index] == '0') return;
      string temp = s.substr(index, i);
      if (stoi(temp) > 255) return;
      path[depth] = temp;
      dfs(s, depth + 1, index + i, path, ans);
      path[depth] = "";
    }
  }
};

File: leetcode-04/solutions/python3/0093.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        ans = []

        self.dfs(s, 0, 0, [""] * 4, ans)

        return ans

    def dfs(self, s: str, depth: int, index: int, path: List[str], ans: List[str]) -> None:
        if depth == 4 and index == len(s):
            ans.append(".".join(path))
            return
        if depth == 4 or index == len(s):
            return

        for i in range(1, 4):
            if index + i > len(s):
                return
            if i > 1 and s[index] == '0':
                return
            temp = s[index: index + i]
            if int(temp) > 255:
                return
            path[depth] = temp
            self.dfs(s, depth + 1, index + i, path, ans)
            path[depth] = ""
